### PR TITLE
Write legacy information to accounts file

### PR DESCRIPTION
**Pull request by ddoktorski**
#1202 

This PR is part of the stack:

-> Write legacy information to accounts file (#1849)
-- Enable using Cairo 1 accounts in sncast (#1850)
-- Use Cairo 1 accounts in sncast tests (#1851)
-- Update documentation and default class hash (#1900)

## Introduced changes

- Write `legacy` information to accounts file in `account create` and `account add` to distinguish between Cairo 0 and Cairo >=1 accounts
- Rename `DEVNET_OZ_CLASS_HASH` -> `DEVNET_OZ_CLASS_HASH_CAIRO_0`

## Checklist

- [X] Linked relevant issue
- [ ] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/sncast/src/lib.rs [hunk 1]
```diff
@@ -48,6 +48,7 @@ struct Account {
     salt: Option<String>,
     deployed: Option<bool>,
     class_hash: Option<String>,
+    legacy: Option<bool>,
```
Comment (THenry14) [inline crates/sncast/src/lib.rs:51]: I'm afraid that the definition of legacy may be too fluid for the future. Maybe we should just hold cairo version in here?

File: crates/sncast/src/lib.rs [hunk 2]
```diff
@@ -327,6 +328,54 @@ fn get_account_from_accounts_file<'a>(
...
+async fn extract_class_hash_or_fetch(
+    class_hash: Option<FieldElement>,
+    address: FieldElement,
+    provider: &JsonRpcClient<HttpTransport>,
+) -> Result<FieldElement> {
+    match class_hash {
+        Some(class_hash) => Ok(class_hash),
+        None => get_class_hash_by_address(provider, address)
+            .await?
+            .ok_or_else(|| anyhow!("No class hash exists for the provided address {address:#x}")),
```
Comment (THenry14) [inline crates/sncast/src/lib.rs:353]: is it possible? :D Maybe we should rephrase the error message to something like "Failed to fetch class hash for address {}` ?

File: crates/sncast/src/lib.rs (hunk 1 above)
Comment (ddoktorski) [reply crates/sncast/src/lib.rs:51]: This way, it aligns with the keystore account files, making it easier to use the same `Account` struct. Moreover, differentiating between Cairo 1 and 2 accounts seems non-trivial (please let me know if there a simple to way to achieve this), and it's not relevant in terms of encoding.

File: crates/sncast/src/lib.rs (hunk 2 above)
Comment (ddoktorski) [reply crates/sncast/src/lib.rs:353]: In this case "No class hash exists" actually means that the provided address does not exist in the given network, so maybe let's rephrase it to `The provided address {} does not exist`?

File: crates/sncast/src/lib.rs (hunk 1 above)
Comment (THenry14) [reply crates/sncast/src/lib.rs:51]: ok, makes sense. guess we will have to wait and see how thing will look like with newer cairo versions then.

File: crates/sncast/src/lib.rs (hunk 2 above)
Comment (ddoktorski) [reply crates/sncast/src/lib.rs]: After all I decided to remove the `extract_class_hash_or_fetch` function in order to limit RPC calls. All RPC errors related to getting the contract class are now being handled by the `handle_rpc_error` function.

File: crates/sncast/src/lib.rs [hunk 3]
```diff
@@ -327,6 +328,42 @@ fn get_account_from_accounts_file<'a>(
...
 }
 
+pub async fn check_if_legacy_contract(
+    class_hash: Option<FieldElement>,
+    address: FieldElement,
+    provider: &JsonRpcClient<HttpTransport>,
+) -> Result<bool> {
+    let contract_class = match class_hash {
+        Some(class_hash) => provider.get_class(BlockId::Tag(Pending), class_hash).await,
+        None => provider.get_class_at(BlockId::Tag(Pending), address).await,
```
Comment (THenry14) [inline crates/sncast/src/lib.rs:338]: should this use `get_class_hash_by_address`?

Comment (ddoktorski) [reply crates/sncast/src/lib.rs:338]: We want to receive a contract class, not a class hash. I initially moved `get_class_hash_by_address` here since it was utilized in the first solution. If `account/add.rs` is a better location for it, it can certainly be moved back.

Comment (THenry14) [reply crates/sncast/src/lib.rs:338]: ahh shit, misread. sorry.

**Review by THenry14 — APPROVED** [review]
